extensions/amp-skimlinks/0.1/tracking.js

```diff
@@ -0,0 +1,290 @@
+ * send the 3 potential Skimlinks tracking requests to Skimlinks tracking API:
+ *  - Page impression tracking
+ *  - Link impression tracking
+ *  - Non-affiliate click tracking.
+ *
+ * It uses the amp-analytics internal API (https://github.com/ampproject/amphtml/blob/master/extensions/amp-analytics/amp-components-analytics.md)
+ * in order to send the tracking requests.
+ */
+export class Tracking {
+  /**
+   * Use tracking instance to track page impressions,
+   * link impressions and non-affiliated clicks.
+   * @param {!AmpElement} element
+   * @param {!Object} skimOptions
+   */
+  constructor(element, skimOptions) {
+    /** @private {boolean} */
+    this.tracking_ = skimOptions.tracking;
+
+    /** @private {!Object} */
+    this.trackingInfo_ = {
+      pubcode: skimOptions.pubcode,
+      // https://github.com/ampproject/amphtml/blob/master/spec/amp-var-substitutions.md
+      pageUrl: 'CANONICAL_URL',
+      referrer: 'DOCUMENT_REFERRER',
+      timezone: 'TIMEZONE',
+      pageImpressionId: generatePageImpressionId(),
+      customTrackingId: skimOptions.customTrackingId,
+      guid: null,
+    };
+
+    /** @private {!Object} */
+    this.skimOptions_ = skimOptions;
+
+    /** @private {!Object} */
+    this.analytics_ = this.setupAnalytics_(element);
+  }
+
+  /**
+   * Getter to access the tracking data from outside the class.
+   * @return {!Object}
+   * @public
+   */
+  getTrackingInfo() {
+    return this.trackingInfo_;
+  }
+
+  /**
+   * Setter to update the tracking data from outside the class.
+   * This is mainly used for setting the guid that we receive from beaconAPI.
+   * @param {!Object} newInfo
+   * @public
+   */
+  setTrackingInfo(newInfo) {
+    Object.assign(this.trackingInfo_, newInfo);
+  }
+
+  /**
+   * Sends "Page impression" and "Link impression" tracking requests (POST).
+   * @param {!./link-rewriter/link-rewriter.AnchorReplacementList} anchorReplacementList
+   * @public
+   */
+  sendImpressionTracking(anchorReplacementList) {
+    if (!this.tracking_) {
+      return;
+    }
+    const {
+      pageImpressionId,
+      timezone,
+      pubcode,
+      pageUrl,
+      guid,
+    } = this.trackingInfo_;
+
+    // This data is common to both page & link impression requests.
+    const commonData = dict({
+      'pub': pubcode,
+      'pag': pageUrl,
+      'guid': guid,
+      'uuid': pageImpressionId,
+      'tz': timezone,
+      'platform': PLATFORM_NAME,
+    });
+
+    const {numberAffiliateLinks, urls} = this.extractAnchorTrackingInfo_(
+        anchorReplacementList
+    );
+
+    this.sendPageImpressionTracking_(
+        commonData,
+        numberAffiliateLinks
+    );
+    this.sendLinkImpressionTracking_(commonData, numberAffiliateLinks, urls);
+  }
+
+  /**
+   * Sends tracking request to Skimlinks tracking API in order to
+   * register non-affiliated click.
+   * @param {!HTMLElement} anchor
+   * @public
+   */
+  sendNaClickTracking(anchor) {
```

Comment:
Out of curiosity: in practice, are these fired in time before page navigation is triggered?

Reply:
By default, the browser doesn't wait. However, we are using the sendBeacon API which has been created for this type of scenarios. https://developer.mozilla.org/en-US/docs/Web/API/Navigator/sendBeacon